### PR TITLE
Support QRDA Cat I R5.2

### PR DESCRIPTION
**Pull request by dczulada**
Adverse Event

Encounter Performed

Allergy Intolerance

null flavor dates

Assessment Performed

Assessment Order

Assessment Recommended

move wrapping tags

pass in xpath

prefix needed

Communication Performed

two flavors of relevant time

Device Applied

Device Order

Device Recommended

Diagnosis

Diagnostic Study Performed

TypeCode

Diagnostic Story Order

DSO template id

DSP template ID

Diagnostic Study Recommended

Medication Discharge

Encounter Order

Encounter Recommended

Family History

Immunization Administered

Immunization Order

Intervention Order

Intervention Performed

Intervention Performed Template Id

Intervention Recommended

Lab Test Performed

Lab Test Order

Lab Test Recommended

Medication Active

Medication Administered

Medication Dispensed

Medication Order

Physical Exam Performed

Physical Exam Order

Physical Exam Requested

Precedure Performed

Procedure Order

Procedure Recommended

Provider Care Experience

Patient Care Experience

Substance Recommended

Symptom

encounter performed

more changes

add related person

ordering of performer in medication administered

update Allergy Intolerance

use hqmfOid for entities

Use cqm-models branch

Updates from QRDA Cat I R5.2 comments

update device applied template id

Pull requests into cqm-parsers require the following. Submitter and reviewer should :white_check_mark: when done. For items that are not-applicable, note it's not-applicable ("N/A") and :white_check_mark:.

**Submitter:**
- [ ] This pull request describes why these changes were made.
- [ ] Internal ticket for this PR:
- [ ] Internal ticket links to this PR
- [ ] Code diff has been done and been reviewed
- [ ] Tests are included and test edge cases
- [ ] Tests have been run locally and pass

**Reviewer 1:**

Name:
- [ ] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [ ] The tests appropriately test the new code, including edge cases
- [ ] You have tried to break the code

**Reviewer 2:**

Name:
- [ ] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [ ] The tests appropriately test the new code, including edge cases
- [ ] You have tried to break the code

### CODE REVIEW

File: lib/qrda-import/data-element-importers/substance_administered_importer.rb [hunk 1]
```diff
@@ -18,6 +19,7 @@ def create_entry(entry_element, nrh = NarrativeReferenceHandler.new)
         substance_administered.dosage = extract_scalar(entry_element, @dosage_xpath)
         substance_administered.frequency = frequency_as_coded_value(entry_element, @frequency_xpath)
         substance_administered.route = code_if_present(entry_element.at_xpath(@route_xpath))
+        substance_administered.performer = extract_entity(entry_element, "./cda:participant[@typeCode='PRF']")
```
Comment (lmd59) [inline lib/qrda-import/data-element-importers/substance_administered_importer.rb]: only substance_recommended???

**Comment by lmd59** [comment]
Other comments:
Medication Dispensed issue should be shared with Paul (extension was not appropriately updated in the new QRDA spec)
Medication Supply Request: update required, contained within medication order
Program Participation: need to add ticket to support program participation
Result: need to translate result observation template to a result template: at least need to update extension and template name/oid, may also do something with result datetime and others
Status: at least need to update extension, may also do something with status datetime

File: lib/qrda-export/catI-r5/qrda_templates/assessment_order.mustache [hunk 2]
```diff
@@ -13,6 +13,12 @@
       <!-- QDM Attribute: Author dateTime -->
       {{> qrda_templates/template_partials/_author}}
     {{/authorDatetime}}
+    {{#requester}}
+      <!-- QDM Attribute: Requester -->
+      <participant typeCode="PRF">
+        {{> qrda_templates/template_partials/_entity}}
+      </participant>
+    {{/requester}}
     {{#reason}}
```
[the remark this answers is not in the document]
Comment (lmd59) [reply lib/qrda-export/catI-r5/qrda_templates/assessment_order.mustache]: Should be shown as a schematron error, may cause issues with test patients if there are any with negations without reasons

File: lib/qrda-export/catI-r5/qrda_templates/assessment_recommended.mustache [hunk 3]
```diff
@@ -13,6 +13,12 @@
       <!-- QDM Attribute: Author dateTime -->
       {{> qrda_templates/template_partials/_author}}
```
[the remark this answers is not in the document]
Comment (lmd59) [reply lib/qrda-export/catI-r5/qrda_templates/assessment_recommended.mustache]: has never been in our model

File: lib/qrda-export/catI-r5/qrda_templates/communication_performed.mustache [hunk 4]
```diff
@@ -11,27 +11,26 @@
...
         {{/recipient}}
         {{#sender}}
         <participant typeCode="AUT">
-            <participantRole>
-                <code {{> _code}}/>
-            </participantRole>
+          <!-- QDM Attribute: Sender -->
+          {{> qrda_templates/template_partials/_entity}}
         </participant>
         {{/sender}}
```
[the remark this answers is not in the document]
Comment (lmd59) [reply lib/qrda-export/catI-r5/qrda_templates/communication_performed.mustache]: Reason template is explicitly allowed (rather than implicitly in text description)

File: lib/qrda-export/catI-r5/qrda_templates/encounter_performed.mustache [hunk 5]
```diff
@@ -34,9 +34,15 @@
...
+        {{#participant}}
+          <!-- QDM Attribute: Participant -->
+          <participant typeCode="PRF">
+            {{> qrda_templates/template_partials/_entity}}
+          </participant>
+        {{/participant}}
         {{#diagnoses}}
           <!-- QDM Attribute: Diagnoses -->
-          {{> qrda_templates/template_partials/_encounter_diagnosis}}
+          {{> qrda_templates/template_partials/_encounter_diagnosis_qdm}}
```
[the remark this answers is not in the document]
Comment (lmd59) [reply lib/qrda-export/catI-r5/qrda_templates/encounter_performed.mustache]: yes

File: lib/qrda-export/catI-r5/qrda_templates/encounter_recommended.mustache [hunk 6]
```diff
@@ -1,15 +1,15 @@
 <entry>
   <act classCode="ACT" moodCode="INT" {{{negation_ind}}}>
     <!--Encounter Recommended Act (V2) -->
-    <templateId root="2.16.840.1.113883.10.20.24.3.134" extension="2017-08-01"/>
+    <templateId root="2.16.840.1.113883.10.20.24.3.134" extension="2019-12-01"/>
```
[the remark this answers is not in the document]
Comment (lmd59) [reply lib/qrda-export/catI-r5/qrda_templates/encounter_recommended.mustache]: have always had id in our template despite the fact it was not a part of the previous QRDA version

File: lib/qrda-export/catI-r5/qrda_templates/immunization_administered.mustache [hunk 7]
```diff
@@ -1,12 +1,16 @@
 <entry>
   <substanceAdministration classCode="SBADM" moodCode="EVN" {{{negation_ind}}}>
     <!-- C-CDA R2 Immunization Activity -->
-    <templateId root="2.16.840.1.113883.10.20.22.4.52" extension="2014-06-09"/>
+    <templateId root="2.16.840.1.113883.10.20.22.4.52" extension="2015-08-09"/>
```
[the remark this answers is not in the document]
Comment (lmd59) [reply lib/qrda-export/catI-r5/qrda_templates/immunization_administered.mustache]: Probably something that we had the wrong extension previously: should be 2015-08-01
This was in the errata update, which was fixed after master had been updated to "support" the new QDM

File: lib/qrda-export/catI-r5/qrda_templates/immunization_administered.mustache [hunk 8]
```diff
@@ -1,12 +1,16 @@
...
     <!-- Immunization Administered -->
-    <templateId root="2.16.840.1.113883.10.20.24.3.140" extension="2017-08-01"/>
+    <templateId root="2.16.840.1.113883.10.20.24.3.140" extension="2019-12-01"/>
     <id root="1.3.6.1.4.1.115" extension="{{object_id}}"/>
     <code code="416118004" codeSystem="2.16.840.1.113883.6.96" codeSystemName="SNOMED CT" displayName="Administration"/>
     <statusCode code="completed"/>
+    {{#relevantDatetime}}
```
[the remark this answers is not in the document]
Comment (lmd59) [reply lib/qrda-export/catI-r5/qrda_templates/immunization_administered.mustache]: author datetime was actually re-added after it was changed into a relevant datetime

[the remark this answers is not in the document]
Comment (lmd59) [reply lib/qrda-export/catI-r5/qrda_templates/immunization_administered.mustache]: Should now be an embedded author

File: lib/qrda-export/catI-r5/qrda_templates/immunization_order.mustache [hunk 9]
```diff
@@ -28,6 +28,12 @@
...
     {{/authorDatetime}}
+    {{#requester}}
+      <!-- QDM Attribute: Requester -->
+      <participant typeCode="PRF">
+        {{> qrda_templates/template_partials/_entity}}
+      </participant>
+    {{/requester}}
     {{#reason}}
       <!-- QDM Attribute: Reason -->
       {{> qrda_templates/template_partials/_reason}}
```
[the remark this answers is not in the document]
Comment (lmd59) [reply lib/qrda-export/catI-r5/qrda_templates/immunization_order.mustache]: Immunization Supply request should be used for supply, and REFR is related
Value is covered by active_datetime method

File: lib/qrda-export/catI-r5/qrda_templates/intervention_performed.mustache [hunk 10]
```diff
@@ -3,19 +3,31 @@
...
+    {{^relevantDatetime}}
+      {{#relevantPeriod}}
+        <!-- QDM Attribute: Relevant Period -->
+        {{{relevant_period}}}
+      {{/relevantPeriod}}
+    {{/relevantDatetime}}
     {{#authorDatetime}}
       <!-- QDM Attribute: Author dateTime -->
       {{> qrda_templates/template_partials/_author}}
     {{/authorDatetime}}
```
[the remark this answers is not in the document]
Comment (lmd59) [reply lib/qrda-export/catI-r5/qrda_templates/intervention_performed.mustache]: no

[the remark this answers is not in the document]
Comment (lmd59) [reply lib/qrda-export/catI-r5/qrda_templates/intervention_performed.mustache]: two different author templates

File: lib/qrda-export/catI-r5/qrda_templates/intervention_recommended.mustache [hunk 11]
```diff
@@ -13,6 +13,12 @@
...
     {{/authorDatetime}}
+    {{#requester}}
+      <!-- QDM Attribute: Requester -->
+      <participant typeCode="PRF">
+        {{> qrda_templates/template_partials/_entity}}
+      </participant>
+    {{/requester}}
     {{#reason}}
       <!-- QDM Attribute: Reason -->
       {{> qrda_templates/template_partials/_reason}}
```
[the remark this answers is not in the document]
Comment (lmd59) [reply lib/qrda-export/catI-r5/qrda_templates/intervention_recommended.mustache]: have always had id

File: lib/qrda-export/catI-r5/qrda_templates/medication_active.mustache [hunk 12]
```diff
@@ -3,15 +3,21 @@
...
+    <templateId root="2.16.840.1.113883.10.20.24.3.41" extension="2019-12-01"/>
 
     <id root="1.3.6.1.4.1.115" extension="{{object_id}}"/>
     <text>{{description}}</text>
     <statusCode code="active"/>
-    {{#relevantPeriod}}
-      <!-- QDM Attribute: Relevant Period -->
-      {{{medication_duration_effective_time}}}
-    {{/relevantPeriod}}
+    {{#relevantDatetime}}
```
[the remark this answers is not in the document]
Comment (lmd59) [reply lib/qrda-export/catI-r5/qrda_templates/medication_active.mustache]: conforms to medication activity, which has that change

File: lib/qrda-export/catI-r5/qrda_templates/medication_administered.mustache [hunk 13]
```diff
@@ -3,14 +3,20 @@
...
-    <templateId root="2.16.840.1.113883.10.20.24.3.42" extension="2017-08-01"/>
+    <templateId root="2.16.840.1.113883.10.20.24.3.42" extension="2019-12-01"/>
     <id root="1.3.6.1.4.1.115" extension="{{object_id}}"/>
     <code code="416118004" codeSystem="2.16.840.1.113883.6.96" codeSystemName="SNOMED CT" displayName="Administration"/>
     <statusCode code="completed"/>
-    {{#relevantPeriod}}
-      <!-- QDM Attribute: Relevant Period -->
-      {{{medication_duration_effective_time}}}
-    {{/relevantPeriod}}
+    {{#relevantDatetime}}
```
[the remark this answers is not in the document]
Comment (lmd59) [reply lib/qrda-export/catI-r5/qrda_templates/medication_administered.mustache]: conforms to medication activity, which has that change

File: lib/qrda-export/catI-r5/qrda_templates/medication_dispensed.mustache [hunk 14]
```diff
@@ -1,23 +1,29 @@
...
+          {{{relevant_date_time_value}}}
+        {{/relevantDatetime}}
+        {{^relevantDatetime}}
+          {{#relevantPeriod}}
+            <!-- QDM Attribute: Relevant Period -->
+            {{{medication_duration_effective_time}}}
+          {{/relevantPeriod}}
+        {{/relevantDatetime}}
         <product>
           <manufacturedProduct classCode="MANU">
```
[the remark this answers is not in the document]
Comment (lmd59) [reply lib/qrda-export/catI-r5/qrda_templates/medication_dispensed.mustache]: Already had product
REFR is conformance statement number change

File: lib/qrda-export/catI-r5/qrda_templates/physical_exam_performed.mustache [hunk 15]
```diff
@@ -3,16 +3,22 @@
...
     <id root="1.3.6.1.4.1.115" extension="{{object_id}}"/>
     <!-- QDM Attribute: Code -->
     {{> _codes}}
     <text>{{description}}</text>
     <statusCode code="completed"/>
-    {{#relevantPeriod}}
-      <!-- QDM Attribute: Relevant Period -->
-      {{{relevant_period}}}
-    {{/relevantPeriod}}
+    {{#relevantDatetime}}
```
[the remark this answers is not in the document]
Comment (lmd59) [reply lib/qrda-export/catI-r5/qrda_templates/physical_exam_performed.mustache]: relevant datetime is usually effective time that has a value

File: lib/qrda-export/catI-r5/qrda_templates/physical_exam_performed.mustache [hunk 16]
```diff
@@ -33,6 +39,12 @@
...
     {{/authorDatetime}}
+    {{#performer}}
+      <!-- QDM Attribute: Performer -->
+      <participant typeCode="PRF">
+        {{> qrda_templates/template_partials/_entity}}
+      </participant>
+    {{/performer}}
     {{#reason}}
       <!-- QDM Attribute: Reason -->
       {{> qrda_templates/template_partials/_reason}}
```
[the remark this answers is not in the document]
Comment (lmd59) [reply lib/qrda-export/catI-r5/qrda_templates/physical_exam_performed.mustache]: same as above

File: lib/qrda-export/catI-r5/qrda_templates/procedure_performed.mustache [hunk 17]
```diff
@@ -1,17 +1,23 @@
...
     <templateId root="2.16.840.1.113883.10.20.22.4.14" extension="2014-06-09"/>
     <id root="1.3.6.1.4.1.115" extension="{{object_id}}"/>
     {{> _codes}}
     <text>{{description}}</text>
     <statusCode code="completed"/>
-    {{#relevantPeriod}}
-      <!-- QDM Attribute: Relevant Period -->
-      {{{relevant_period}}}
-    {{/relevantPeriod}}
+    {{#relevantDatetime}}
```
[the remark this answers is not in the document]
Comment (lmd59) [reply lib/qrda-export/catI-r5/qrda_templates/procedure_performed.mustache]: see above

File: lib/qrda-export/catI-r5/qrda_templates/procedure_recommended.mustache [hunk 18]
```diff
@@ -17,6 +17,12 @@
       <!-- QDM Attribute: Author dateTime -->
       {{> qrda_templates/template_partials/_author}}
     {{/authorDatetime}}
+    {{#requester}}
+      <!-- QDM Attribute: Requester -->
+      <participant typeCode="PRF">
```
[the remark this answers is not in the document]
Comment (lmd59) [reply lib/qrda-export/catI-r5/qrda_templates/procedure_recommended.mustache]: mistake in QRDA

File: lib/qrda-export/catI-r5/qrda_templates/procedure_performed.mustache [hunk 19]
```diff
@@ -24,6 +30,16 @@
...
+      <!-- QDM Attribute: Performer -->
+      <participant typeCode="PRF">
+        {{> qrda_templates/template_partials/_entity}}
+      </participant>
+    {{/performer}}
+    {{#rank}}
+      <!-- QDM Attribute: Rank -->
+      {{> qrda_templates/template_partials/_rank}}
+    {{/rank}}
     {{#reason}}
```
[the remark this answers is not in the document]
Comment (dczulada) [reply lib/qrda-export/catI-r5/qrda_templates/procedure_performed.mustache]: This was related to the prior blanket statement that all entries can have a reason for negation, now it's explicit.

File: lib/qrda-export/catI-r5/qrda_templates/procedure_recommended.mustache [hunk 20]
```diff
@@ -25,6 +31,10 @@
       <!-- QDM Attribute: Negation Rationale -->
       {{> qrda_templates/template_partials/_reason}}
     {{/negationRationale}}
+    {{#rank}}
+      <!-- QDM Attribute: Rank -->
+      {{> qrda_templates/template_partials/_rank}}
+    {{/rank}}
   </procedure>
 </entry>
```
[the remark this answers is not in the document]
Comment (dczulada) [reply lib/qrda-export/catI-r5/qrda_templates/procedure_recommended.mustache]: yes, that is in the rank partial.
